File: Stack/Evaluate_Reverse_Polish_Notation/3_Optimal.java
import java.util.*;

class Solution {
    public int evalRPN(String[] tokens) {
        // Use a plain int array as a stack to avoid boxing overhead.
        int[] stack = new int[tokens.length];
        int top = 0; // next free position
        for (String tok : tokens) {
            if (tok.length() == 1 && (tok.charAt(0) == '+' || tok.charAt(0) == '-' || tok.charAt(0) == '*' || tok.charAt(0) == '/')) {
                int b = stack[--top];
                int a = stack[--top];
                int res;
                switch (tok.charAt(0)) {
                    case '+': res = a + b; break;
                    case '-': res = a - b; break;
                    case '*': res = a * b; break;
                    default:  res = a / b; break; // '/'
                }
                stack[top++] = res;
            } else {
                stack[top++] = Integer.parseInt(tok);
            }
        }
        return stack[0];
    }
}

File: Stack/Evaluate_Reverse_Polish_Notation/3_Optimal.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        // Pre‑allocate an array to act as a stack.
        vector<int> stk(tokens.size());
        int top = 0; // points to next free slot
        for (const string &tok : tokens) {
            if (tok.size() == 1 && (tok[0] == '+' || tok[0] == '-' || tok[0] == '*' || tok[0] == '/')) {
                int b = stk[--top];
                int a = stk[--top];
                int res = 0;
                switch (tok[0]) {
                    case '+': res = a + b; break;
                    case '-': res = a - b; break;
                    case '*': res = a * b; break;
                    case '/': res = a / b; break; // truncates toward zero
                }
                stk[top++] = res;
            } else {
                stk[top++] = stoi(tok);
            }
        }
        return stk[0];
    }
};
